_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,280 @@
+was born out of Google, which historically wrote code for homogeneous fleets of
+Linux servers with little need for customization. What customization was needed
+was achieved with ad hoc flags and logic built
+[straight](https://source.bazel.build/bazel/+/144912e7b7a86b45e07f79e76f6fed20890acb36:src/main/java/com/google/devtools/build/lib/rules/cpp/CppOptions.java;l=258)
+into the build tool.
+
+This is no longer true inside or outside of Google. Modern software targets
+phones, cloud, servers, desktops, [smart
+devices](https://en.wikipedia.org/wiki/Smart_device) and more, and must offer
+increasingly flexible support for any intended combination of these platforms.
+
+This means
+
+    $ bazel build //my:binary
+
+isn't enough to describe what you want to do. What platform(s) are you
+targeting? What features do you want to include? What if you want to build a
+complex app with generated sources, client and server modules, native
+extensions, and test data?
+
+Bazel's historical approach of ad hoc flags:
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+is lacking in many ways:
+
+1. `--cpu` only accepts values explicitly supported by C++ rules. Even if you're
+   not building C++.
+1. `--crosstool_top`, which specifies exactly what command line should compile
+   your code, is even more tied to C++. Other languages may be organized
+   completely differently.
+1. `--define` is completely unstructured and unreadable in Starlark.
+1. `$ bazel build //my:all` sets the same flags for all targets, even if
+   different targets need different flags.
+1. `//my:binary`'s deps must use the same settings as their parent.
+1. It might be hard to remember which flags are needed by which targets,
+   especially across users and dev/test/prod environments. This risks
+   inconsistent and possibly incorrect builds.
+
+A different approach might be to set flags directly in targets that need them,
+let rule writers design the flags that affect their rules, and provide standard
+extensible APIs for cross-language concepts like *platform* and *cpu*. This
+lets people configure their projects on their own terms, using language
+consistent with how their projects are organized.
+
+Bazel's configurability work is redesigning Bazel around these ideas. It's
+ultimate goal is to make
+
+    $ bazel build //:all
+
+*just work*, which means every target knows how to build "the right way", for
+whatever platform(s) and features you care about, and you only have to set
+flags that are genuinely interesting to you. All while keeping builds fast and
+correct.
+
+## Configuration vs. Attributes
+
+Rules have always had [attributes](), which also affect how they build. The
+difference between configuration and attributes is that attributes only affect
+the rule's [direct build actions](https://docs.bazel.build/versions/master/skylark/rules.html#implementation-function)).
```

Comment:
Is there a simpler way of explaining this difference? Or perhaps frame this section for just one audience: "If you're thinking to yourself: 'Why not use rule attributes to solve this problem?', that's a great question. ... ..."

This will let anyone who doesn't have that question skip or skim this section.

Reply:
I tried the latter. WDYT?

I can't think of any content in this section that I think is safe to remove. I included this because this is a *very* common source of confusion that makes it hard for people just getting up to speed to reason about the problem on a basic level. And that key difference of "down the graph" vs. "up the graph" I think is a crucial one worth calling out.